lib/stats/ManifestPlugin.js

```diff
@@ -0,0 +1,180 @@
+		validate(options);
+
+		/** @type {Required<ManifestPluginOptions>} */
+		this.options = {
+			filename: "manifest.json",
+			handle: (manifest, _stats) => JSON.stringify(manifest, null, 2),
+			...options
+		};
+	}
+
+	/**
+	 * Apply the plugin
+	 * @param {Compiler} compiler the compiler instance
+	 * @returns {void}
+	 */
+	apply(compiler) {
+		/** @type {WeakMap<Compilation, StatsCompilation>} */
+		const cachedStats = new WeakMap();
+
+		compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation) => {
+			compilation.hooks.processAssets.tap(
+				{
+					name: PLUGIN_NAME,
+					stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE
+				},
+				() => {
+					let stats =
+						/** @type {StatsCompilation | undefined} */ cachedStats.get(
+							compilation
+						);
+					if (!stats) {
+						stats = compilation.getStats().toJson({
+							all: false,
+							assets: true,
+							cachedAssets: true,
+							assetsSpace: Infinity,
+							publicPath: true
+						});
+						cachedStats.set(compilation, stats);
+					}
+
+					/** @type {Set<string>} */
+					const added = new Set();
+					/**
+					 * @type {{name: string, file: string}[]}
+					 */
+					const items = [];
+
+					/**
+					 * @param {string} file file
+					 * @param {((file: string) => string)=} namer namer
+					 * @returns {void}
+					 */
+					const handleFile = (file, namer) => {
+						if (added.has(file)) return;
+						added.add(file);
+
+						let name = namer ? namer(file) : file;
+						const asset = compilation.getAsset(file);
+						if (asset && asset.info.sourceFilename) {
+							name = path.join(
+								path.dirname(file),
+								path.basename(asset.info.sourceFilename)
+							);
+						}
+						items.push({ name, file });
+					};
+
+					for (const chunk of compilation.chunks) {
+						if (chunk instanceof HotUpdateChunk) continue;
+
+						const chunkName = chunk.name;
+						for (const auxiliaryFile of chunk.auxiliaryFiles) {
+							handleFile(auxiliaryFile, (file) => path.basename(file));
+						}
+						for (const file of chunk.files) {
+							handleFile(file, (file) => {
+								if (chunkName) return `${chunkName}.${extname(file)}`;
+								return file;
+							});
+						}
+					}
+
+					if (stats.assets) {
+						for (const asset of stats.assets) {
+							if (asset.info.hotModuleReplacement) {
+								continue;
+							}
+							handleFile(asset.name);
+						}
+					}
+
+					/** @type {Record<string, string>} */
+					const manifest = {};
+					const hashDigestLength = compilation.outputOptions.hashDigestLength;
+
+					/**
+					 * @param {string} name name
+					 * @returns {string} hash removed name
+					 */
+					const removeHash = (name) => {
+						// Handles hashes that match configured `hashDigestLength`
+						// i.e. index.XXXX.html -> index.html (html-webpack-plugin)
+						if (hashDigestLength <= 0) return name;
+						const reg = new RegExp(
+							`(\\.[a-f0-9]{${hashDigestLength}})(?=\\.)`,
+							"gi"
+						);
```

Comment:
When we need to use regexp except `html-webpack-plugin`? Only for entries/chunks? Because for assets we use `sourceFilename`, maybe we can emit for entries/chunks - `originalModuleFilename` property in asset info and use it here to avoid using hashes?